slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/InsertTest.scala

```diff
@@ -230,4 +230,90 @@ class InsertTest extends AsyncTest[JdbcTestDB] {
+      _ <- ts1.insertOrUpdate((1)).map(_ shouldBe 1)
+      _ <- ts2.schema.create
+      _ <- ts2 ++= Seq(V(1, "a"), V(2, "b"))
+      _ <- ts2.insertOrUpdate(V(0, "c")).map(_ shouldBe 1)
+      _ <- ts2.insertOrUpdate(V(1, "a")).map(_ shouldBe 1)
+      _ <- ts2.insertOrUpdate(V(2, "d")).map(_ shouldBe 1)
+    } yield ()).withPinnedSession
+  } else DBIO.seq()
+
+  def testInsertOrUpdateWithIntegrityError: DBIOAction[Unit, NoStream, Effect.All] = {
+    case class Book(id: Long, title: String, code: Option[String] = None)
+    class BooksTable(tag: Tag) extends Table[Book](tag, "books") {
+      def id = column[Long]("id", O.PrimaryKey)
+      def title = column[String]("title", O.SqlType("varchar(20)"))
+      def code = column[Option[String]]("code", O.SqlType("varchar(20)"))
+      override def * = (id, title, code) <> ((Book.apply _).tupled, Book.unapply)
+    }
+    val books = TableQuery[BooksTable]
+    case class BookMeta(id: Long, bookId: Long, tag: Long)
+    class BookMetaTable(tag: Tag) extends Table[BookMeta](tag, "book_meta") {
+      def id    = column[Long]("id")
+      def bookId = column[Long]("book_id")
+      def bookTag  = column[Long]("tag")
+      def pk = primaryKey("book_meta_pk", (bookId, id))
+      def book = foreignKey("book_fk", bookId, books)(_.id, onUpdate=ForeignKeyAction.Restrict, onDelete=ForeignKeyAction.Cascade)
+      override def * = (bookId, id, bookTag) <> ((BookMeta.apply _).tupled, BookMeta.unapply)
+    }
+    val meta = TableQuery[BookMetaTable]
+
+    import scala.util.{Success, Failure}
+    val data = BookMeta(-1, -1, 0)// fail, because there are no books.
+    DBIO.seq(
+      (meta += data).asTry.map {
+        case Success(_) => throw new Exception("Insertion should be failed.")
+        case _ => ()
+      }
+    )
+  }
+
+  def testInsertAndUpdateShouldNotTruncateData = {
+    case class Test(id: Int, name: String)
+    class TestTable(tag: Tag) extends Table[Test](tag, "test") {
+      def id = column[Int]("id", O.PrimaryKey)
+      def name = column[String]("name", O.SqlType("varchar(2)"))
+      def * = (id, name) <> (Test.tupled, Test.unapply)
+    }
+    val ts = TableQuery[TestTable]
+    import scala.util.{Success, Failure}
+    DBIO.seq(
+      ts.schema.create,
+      ts ++= Seq(Test(1, "a")),
+      // sqlite donesn't care this case
+      (ts ++= Seq(Test(2, "123"))).asTry.map {
+        case Success(_) if !tdb.confName.contains("sqlite") => throw new Exception("Data is truncated. It shouldn't be succeeded.")
+        case _ => ()
+      },
+      ts.filter(_.id === 1).map(_.name).update("123").asTry.map {
+        case Success(_) if !tdb.confName.contains("sqlite") => throw new Exception("Data is truncated. It shouldn't be succeeded.")
+        case _ => ()
+      },
+      ts.insertOrUpdate(Test(1, "123")).asTry.map {
+        case Success(_) if !tdb.confName.contains("sqlite") => throw new Exception("Data is truncated. It shouldn't be succeeded.")
+        case _ => ()
+      }
```

Comment:
If SQLite doesn't work with any of the cases in this test then it would be better to add single if on top of the method, like you've done in `def testInsertOrUpdateWithPrimaryKeyOnly`

Reply:
Fixed in https://github.com/slick/slick/pull/2201/commits/dbc35c8722314bfe082aa4fa68301cece352ebce#diff-698cce732005a091c7b0b91de1a124d4c57209104ce2c00b954f93ec46a1a591R294-R295